### PR TITLE
PA Common Alignment Producers 81X

### PR DESCRIPTION
**Pull request by jrcastle**
Adds in new PA-specific common alignment producers for muon-based tracker alignment collections.  In pA and AA collisions, the yields of isoMus and Z's are strongly suppressed by the *RelCombIsoMuons step.  Studies on 2013 pPb data show improvement in alignment yields using producers where this step has been removed as outlined in the following presentation at the tracker alignment weekly meeting:

https://indico.cern.ch/event/536888/contributions/2326743/attachments/1348506/2034547/10_5_16_Alignment.pdf

Additional edits were made to the new Millepede configuration files to include these new track collections in the standard workflows.

### CODE REVIEW

**Comment by jrcastle** [comment]
@mandrenguyen

**Comment by mandrenguyen** [comment]
@fabozzi  Alca has requested that these pA-specific producers be tested as part of the pA relval, when it's updated.  @jrcastle what are the cmsDriver options needed to activate these workflows?

**Comment by jrcastle** [comment]
-s ALCA:TkAlMuonIsolatedPA
-s ALCA:TkAlUpsilonMuMuPA
-s ALCA:TkAlZMuMuPA

**Comment by mmusich** [comment]
please test

**Comment by mmusich** [comment]
> @fabozzi Alca has requested that these pA-specific producers be tested as part of the pA relval, when it's updated. @jrcastle what are the cmsDriver options needed to activate these workflows?

@fabozzi: once the corresponding wf in the matrix is available would it be possible to advertise it? 

@mandrenguyen: the pA relval you mention will be "typical" in terms of event topology and beam conditions, with respect to what is foreseen for the 2016 pA run? we ought to measure the new AlCaReco throughput before activating the new worfklows at tier-0

File: Configuration/EventContent/python/AlCaRecoOutput_cff.py [hunk 1]
```diff
@@ -17,14 +18,14 @@
 # AlCaReco for track based alignment using isoMu events
 from Alignment.CommonAlignmentProducer.ALCARECOTkAlMuonIsolated_Output_cff import *
 from Alignment.CommonAlignmentProducer.ALCARECOTkAlMuonIsolatedHI_Output_cff import *
-# AlCaReco for track based alignment using isoMu events for PA data-taking
```
Comment (mmusich) [inline Configuration/EventContent/python/AlCaRecoOutput_cff.py]: why removing this line?

File: Configuration/StandardSequences/python/AlCaRecoStreams_cff.py [hunk 2]
```diff
@@ -17,12 +18,12 @@
 from Alignment.CommonAlignmentProducer.ALCARECOTkAlCosmics0THLT_cff import *
 # AlCaReco for track based alignment using isoMu events
 from Alignment.CommonAlignmentProducer.ALCARECOTkAlMuonIsolated_cff import *
-# AlCaReco for track based alignment using isoMu events for PA data-taking
```
Comment (mmusich) [inline Configuration/StandardSequences/python/AlCaRecoStreams_cff.py]: ditto

File: Configuration/StandardSequences/python/AlCaRecoStreams_cff.py [hunk 3]
```diff
@@ -272,6 +275,15 @@
 	dataTier = cms.untracked.string('ALCARECO')
 	)
 
+ALCARECOStreamTkAlZMuMuPA = cms.FilteredStream(
+        responsible = 'James Castle',
```
Comment (mmusich) [inline Configuration/StandardSequences/python/AlCaRecoStreams_cff.py]: suggestion, may be you can replace the ownership of ALCARECOStreamTkAlMuonIsolatedPA as well

Comment (jrcastle) [inline Configuration/StandardSequences/python/AlCaRecoStreams_cff.py]: Will do

File: Configuration/StandardSequences/python/AlCaRecoStreams_cff.py (hunk 2 above)
Comment (jrcastle) [inline Configuration/StandardSequences/python/AlCaRecoStreams_cff.py]: It was to put the three ALCARECOs under the same umbrella of isoMu events.  I'll go back and put in comments for the PA producers as well as the HI ones.

File: Alignment/CommonAlignmentProducer/python/ALCARECOTkAlZMuMuPA_cff.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+from Alignment.CommonAlignmentProducer.ALCARECOTkAlZMuMu_cff import *
+
+ALCARECOTkAlZMuMuPAHLT = ALCARECOTkAlZMuMuHLT.clone(
+    eventSetupPathsKey = 'TkAlZMuMu'
+)
+
+ALCARECOTkAlZMuMuPADCSFilter = ALCARECOTkAlZMuMuDCSFilter.clone()
+
+ALCARECOTkAlZMuMuPAGoodMuons = ALCARECOTkAlZMuMuGoodMuons.clone()
+#ALCARECOTkAlZMuMuPARelCombIsoMuons = ALCARECOTkAlZMuMuRelCombIsoMuons.clone()
```
Comment (jrcastle) [inline Alignment/CommonAlignmentProducer/python/ALCARECOTkAlZMuMuPA_cff.py]: On it.

File: Alignment/CommonAlignmentProducer/python/ALCARECOTkAlMuonIsolatedHI_cff.py [hunk 5]
```diff
@@ -1,27 +1,21 @@
-# AlCaReco for track based alignment using isolated muon tracks
+# AlCaReco for track based alignment using isolated muon tracks - relaxed cuts for PbPb collisions
 import FWCore.ParameterSet.Config as cms
 
 from Alignment.CommonAlignmentProducer.ALCARECOTkAlMuonIsolated_cff import *
 
 ALCARECOTkAlMuonIsolatedHIHLT = ALCARECOTkAlMuonIsolatedHLT.clone(
-    eventSetupPathsKey = 'TkAlMuonIsolated'
+    eventSetupPathsKey = 'TkAlMuonIsolatedHI'
     )
```
Comment (jrcastle) [inline Alignment/CommonAlignmentProducer/python/ALCARECOTkAlMuonIsolatedHI_cff.py]: We are, I checked by doing:

runTheMatrix.py -l 140.51

The outcome of which was:

140.51_RunHI2010+RunHI2010+REPACKHID+RECOHID11St3+HARVESTDHI Step0-PASSED Step1-PASSED Step2-PASSED Step3-PASSED  - time date Thu Oct  6 17:49:41 2016-date Thu Oct  6 17:45:09 2016; exit: 0 0 0 0
1 1 1 1 tests passed, 0 0 0 0 failed

Comment (mmusich) [inline Alignment/CommonAlignmentProducer/python/ALCARECOTkAlMuonIsolatedHI_cff.py]: great, thanks!

**Comment by mmusich** [comment]
@cmsbuild please test

**Comment by ghellwig** [comment]
@jrcastle I am fine with the MillePede-related changes
@mmusich already commented on the other parts or is something missing?

**Comment by jrcastle** [comment]
@ghellwig as far as I'm aware we're now just waiting for signatures.  

@mandrenguyen  @mandrenguyen I see that #16119 has merged.  Will we need to open up a new PR to include the new PA producers in the pPb relval workflow? When they are part we can use that information to address the T0 storage issue that @ghellwig brought up at the alignment DPG meeting. This is what I would like the workflow (listed individually, but will ultimately become part of a single step):
-s ALCA:TkAlMinBias
-s ALCA:TkAlMuonIsolatedPA
-s ALCA:TkAlUpsilonMuMuPA
-s ALCA:TkAlZMuMuPA

**Comment by mmusich** [comment]
Hi @jrcastle please wait PR #16119 to be included in the next IB.
At that point you can edit the  wf. 281 and add your producers in the `ALCA` step

**Comment by jrcastle** [comment]
@mmusich, @mandrenguyen, looking over the new pPb workflow, it seems like the best place to put in the ALCA step would be in RECOUP15_PPb:

`steps['RECOUP15_PPb']=merge([{'-s':'RAW2DIGI,L1Reco,RECO,EI,VALIDATION,DQM','--conditions':'auto:run2_mc_pa','--customise':'RecoHI/Configuration/customise_PPwithHI.customisePPrecoforPPb','--datatier':'AODSIM,DQMIO','--eventcontent':'AODSIM,DQM'}, steps['RECOUP15']])`

I would make the change to:

`steps['RECOUP15_PPb']=merge([{'-s':'RAW2DIGI,L1Reco,RECO,ALCA:TkAlMinBias+TkAlMuonIsolatedPA+TkAlUpsilonMuMuPA+TkAlZMuMuPA,EI,VALIDATION,DQM','--conditions':'auto:run2_mc_pa','--customise':'RecoHI/Configuration/customise_PPwithHI.customisePPrecoforPPb','--datatier':'AODSIM,DQMIO','--eventcontent':'AODSIM,DQM'}, steps['RECOUP15']])`

How does this sound?

**Comment by fabozzi** [comment]
@jrcastle 
Yes, the Alca step should be inserted in RECOUP15_PPb

**Comment by jrcastle** [comment]
@fabozzi, since my branch is now behind the merge, what would you suggest the best way to proceed is then?  Should I cherry pick the commits from #16119 to my current branch, make my changes, and commit them, or should I close this PR and open up a new one (starting from the latest IB) incorporating the new producers as well as the wf?

**Comment by fabozzi** [comment]
@jrcastle 
Actually the changes in the matrix are well localized, so you can try the option to cherry pick the commits from #16119. However, if you have conflicts in merging, then starting from the latest IB is the cleanest option for sure.

**Comment by fabozzi** [comment]
@jrcastle 
it seems there are conflicts...

**Comment by jrcastle** [comment]
Yeah, I've managed to screw this up.  I'll close this and open up a a new one from a fresh IB.

**Pull request closed without merging**